### PR TITLE
Add `ui.MarkdownStream`

### PR DESCRIPTION
**Pull request by cpsievert**
This PR extracts the streaming markdown capabilities of `ui.Chat()` into another component: `ui.MarkdownStream()`. See the new files for example usage:

* `shiny/api-examples/MarkdownStream/app-express.py`
* `shiny/api-examples/MarkdownStream/app-core.py`

While extracting out the markdown streaming portion, I also improved the 'auto-scroll' behavior. Now it should be easier to disable/enable auto scroll by scrolling away/towards the bottom of a scrollable parent of the stream. Also, this is done in such a way that `MarkdownStream()` will automatically find the nearest scrollable parent and scroll that element (so that, it just works in something like a card with a set height).

## Follow up

- [ ] Handle fallout for `shinychat` https://github.com/posit-dev/shinychat/pull/23

### CODE REVIEW

File: shiny/api-examples/MarkdownStream/app-core.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    readme_chunks = readme.text.replace("\n", " \n ").split(" ")
+
+    # Generate words from the README.md file (with a small delay)
+    def chunk_generator():
+        for chunk in readme_chunks:
+            time.sleep(0.05)
+            yield chunk + " "
+
+    md = ui.MarkdownStream("shiny-readme")
+    md.ui()
```
Comment (jcheng5) [inline shiny/api-examples/MarkdownStream/app-core.py:26]:
```suggestion
```

File: shiny/api-examples/MarkdownStream/app-core.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import time
+
+import requests
+
+from shiny import App, ui
+
+app_ui = ui.page_fluid(
+    ui.markdown_stream_ui("shiny-readme"),
```
Comment (jcheng5) [inline shiny/api-examples/MarkdownStream/app-core.py:8]: Would `ui.output_markdown_stream` be more consistent with our past naming, perhaps? (I think of Chat as a special case because it's equal parts output and input)

File: js/chat/chat.ts [hunk 3]
```diff
@@ -492,8 +356,21 @@ $(function () {
...
       const el = document.getElementById(message.id);
-      el?.dispatchEvent(evt);
+
+      if (!el) {
+        const errMsg = `
+          Unable to handle Chat() message since element with id ${message.id} wasn't
+          found. Do you need to call .ui() (Express) or need a
+          chat_ui('${message.id}') in the UI (Core)?
+        `;
+        console.error(errMsg);
```
Comment (cpsievert) [inline js/chat/chat.ts:368]:
```suggestion
```

File: js/markdown-stream/markdown-stream.ts [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+function handleMessage(message: ContentMessage | IsStreamingMessage): void {
+  const el = document.getElementById(message.id) as MarkdownElement;
+
+  if (!el) {
+    const errMsg = `
+      Unable to handle MarkdownStream() message since element with id ${message.id}
+      wasn't found. Do you need to call .ui() (Express) or need a
+      output_markdown_stream('${message.id}') in the UI (Core)?
+    `;
+    console.error(errMsg);
```
Comment (cpsievert) [inline js/markdown-stream/markdown-stream.ts:221]:
```suggestion
```

**Comment by cpsievert** [comment]
@gadenbuie in case you're curious, the current failures are due to the airmass example (i.e., this makes CI no worse than it currently is).

This PR is in a pretty complete state now

**Comment by cpsievert** [comment]
Oh shoot, I forgot about error handling

File: shiny/ui/_markdown_stream.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    def _send_custom_message(self, msg: Union[ContentMessage, isStreamingMessage]):
+        if self._session.is_stub_session():
+            return
+        self._session._send_message_sync(
+            {"custom": {"shinyMarkdownStreamMessage": msg}}
+        )
+
+
+@add_example()
+def output_markdown_stream(
```
Comment (cpsievert) [inline shiny/ui/_markdown_stream.py:194]: FWIW, I feel a bit torn on whether this should be `output_markdown_stream()` or `markdown_stream_ui()`

Comment (cpsievert) [reply shiny/ui/_markdown_stream.py]: [3b4a5a4](https://github.com/posit-dev/py-shiny/pull/1782/commits/3b4a5a4f7e70ffcd028a5f8229779496bc66ff2b) does this renaming

**Comment by cpsievert** [comment]
Ughh, I also forgot that, in order to have a non-blocking stream, the websocket messages need to be asyncronous. This is precisely why `chat.append_message_stream()` is async.

I could maybe be talked having both a `.stream()` and `.stream_async()`, but I'm leaning to just having one async `.stream()` since it's probably not clear to users how much better the async is for concurrent users. It's also consistent with `chat.append_message_stream()`. Unfortunately, `chatlas` has both `.stream()` and `.stream_async()`, but I think I can live with that.

Anyway, [cf36d18](https://github.com/posit-dev/py-shiny/pull/1782/commits/cf36d18518cca66661a04e4b8c7d241ed34e969f) makes the switch from sync to async

File: js/markdown-stream/markdown-stream.scss [hunk 6]
```diff
@@ -0,0 +1,51 @@
+@use "highlight_styles" as highlight_styles;
+
+/* Code highlighting (for both light and dark mode) */
+@include highlight_styles.atom_one_light;
+[data-bs-theme="dark"] {
+  @include highlight_styles.atom_one_dark;
+}
```
Comment (gadenbuie) [inline js/markdown-stream/markdown-stream.scss:7]: Not something to handle in this PR, but we should anticipate highlighting styles being something that people will want to configure.

File: shiny/ui/_markdown_stream.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+@add_example()
+class MarkdownStream:
+    """
+    Stream markdown (or HTML) content.
+
+    Parameters
+    ----------
+    id
+        A unique identifier for this markdown stream.
```
Comment (gadenbuie) [inline shiny/ui/_markdown_stream.py:47]: The use of "markdown stream" here gave me pause. I think it should either be a proper noun, like `MarkdownStream`, or it should name more concretely what it identifies. For me, the confusion is around does "markdown stream" refer to the stream of markdown text itself or the UI component where it will be displayed?

File: shiny/ui/_markdown_stream.py [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+
+class isStreamingMessage(TypedDict):
+    id: str
+    isStreaming: bool
+
+
+@add_example()
+class MarkdownStream:
+    """
+    Stream markdown (or HTML) content.
```
Comment (gadenbuie) [inline shiny/ui/_markdown_stream.py:42]:
```suggestion
    A component for streaming markdown or HTML content.
```

Trying to avoid noun/verb confusion around "stream". To me, a class describes a _noun_ but _stream_ here is used as a verb.

File: shiny/ui/_markdown_stream.py [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+        self,
+        *,
+        content: str = "",
+        content_type: StreamingContentType = "markdown",
+        auto_scroll: bool = True,
+        width: CssUnit = "100%",
+        height: CssUnit = "auto",
+    ) -> Tag:
+        """
+        Get the UI element for this markdown stream.
```
Comment (gadenbuie) [inline shiny/ui/_markdown_stream.py:95]:
```suggestion
        Create a UI element for this markdown stream.
```

File: shiny/ui/_markdown_stream.py [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+
+        This method is only relevant for Shiny Express. In Shiny Core, use
+        :func:`~shiny.ui.output_markdown_stream` for placing the markdown stream
+        in the UI.
```
Comment (gadenbuie) [inline shiny/ui/_markdown_stream.py:99]: Would the Shiny Express pattern work in Core?

```python
md = ui.MarkdownStream("readme")

ui = ui.page_fluid(
    md.ui()
)

def server(input, output, session):
    # Read in the README.md file from the py-shiny repository
    # ... snip ...

    @reactive.effect
    async def _():
        await md.stream(chunk_generator())


app = App(app_ui, server)
```

Comment (cpsievert) [reply shiny/ui/_markdown_stream.py:99]: No, I don't think we want to support that since, if there are multiple user sessions, they would be mutating the same object

File: shiny/ui/_markdown_stream.py [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+            height=height,
+        )
+
+    async def stream(
+        self,
+        content: Iterable[str] | AsyncIterable[str],
+        clear: bool = True,
+    ):
+        """
+        Stream content into the markdown stream.
```
Comment (gadenbuie) [inline shiny/ui/_markdown_stream.py:139]: To touch on "markdown stream" one more time, I don't think it's a great term to describe the UI or component that's in the app. One option could be to just capitalize "Markdown Stream". Otherwise, markdown stream here seems ephemeral and it's not immediately obvious that this means "stream into the content into the UI".

Comment (gadenbuie) [reply shiny/ui/_markdown_stream.py:139]: I'm going to stop calling this out; I think it's worth doing a quick review of all uses.

File: shiny/ui/_markdown_stream.py [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+            auto_scroll=auto_scroll,
+            width=width,
+            height=height,
+        )
+
+    async def stream(
+        self,
+        content: Iterable[str] | AsyncIterable[str],
+        clear: bool = True,
+    ):
```
Comment (gadenbuie) [inline shiny/ui/_markdown_stream.py:137]: I think it's worth considering having either a `.set()` method (or better named) or at least including in the docs that `.stream([content])` can be used to set the content. Primarily because we allow setting the initial content and you'll likely want to be able to restore that initial state if you called the `.clear()` method.

Relatedly, I can see utility in having an `.append()` that you could use synchronously. AFAICT that's entirely supported by the component but would be hard to access as a user.

File: shiny/ui/_markdown_stream.py [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+    id
+        A unique identifier for this markdown stream.
+    content
+        Some content to display before any streaming occurs.
+    content_type
+        The content type. Default is "markdown" (specifically, CommonMark).
+        Other supported options are:
+        - `"html"`: for rendering HTML content.
+        - `"text"`: for plain text.
+        - `"semi-markdown"`: for rendering markdown, but with HTML tags escaped.
```
Comment (gadenbuie) [inline shiny/ui/_markdown_stream.py:254]: Is it possible to change this choice after you've created the UI element? It's worth considering.

Or maybe it's that this feels like it's in the wrong place, since the message is the thing that's in html, text or markdown. I can see this being less relevant for the chat use case but as a standalone component it seems feasible to want to switch between html/text/markdown.

I think we should leave `content_type` here, to pair with `content`, but we should also have `content_type` as an argument in `stream()`. That argument could default to `None` – i.e. don't change the content type – but if provided we'd change the content type with the first message. (Just brainstorming out loud...)

File: js/markdown-stream/markdown-stream.ts [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+    this.#scrollableElement?.removeEventListener("scroll", this.#onScroll);
+    this.#scrollableElement = null;
+    this.#isUserScrolled = false;
+  }
+}
+
+// ------- Register custom elements and shiny bindings ---------
+
+customElements.get("shiny-markdown-stream") ||
+  customElements.define("shiny-markdown-stream", MarkdownElement);
```
Comment (gadenbuie) [inline js/markdown-stream/markdown-stream.ts:223]:
```suggestion
if (!customElements.get("shiny-markdown-stream")) {
  customElements.define("shiny-markdown-stream", MarkdownElement);
}
```

Minor nit: I find this version easier to read while skimming, especially since the long expression is wrapped.

File: shiny/ui/_markdown_stream.py [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+        if self.on_error == "unhandled":
+            raise e
+        else:
+            sanitize = self.on_error == "sanitize"
+            msg = f"Error in MarkdownStream('{self.id}'): {str(e)}"
+            raise NotifyException(msg, sanitize=sanitize) from e
+
+
+@add_example()
+def markdown_stream_ui(
```
Comment (gadenbuie) [inline shiny/ui/_markdown_stream.py:228]: Could we call this `output_markdown_stream()`? That feels like a better fit, or at least to fit with most user expectations. I think I'd rather `ui.output_markdown_stream()` than `ui.markdown_stream_ui()`.

File: js/markdown-stream/markdown-stream.ts [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+
+function handleMessage(message: ContentMessage | IsStreamingMessage): void {
+  const el = document.getElementById(message.id) as MarkdownElement;
+
+  if (!el) {
+    showShinyClientMessage({
+      status: "error",
+      message: `Unable to handle MarkdownStream() message since element with id
+      ${message.id} wasn't found. Do you need to call .ui() (Express) or need a
+      output_markdown_stream('${message.id}') in the UI (Core)?`,
```
Comment (gadenbuie) [inline js/markdown-stream/markdown-stream.ts:233]: It's currently

```suggestion
      markdown_stream_ui('${message.id}') in the UI (Core)?`,
```

but I like `output_markdown_stream()` better (unless you've already considered and moved away from this for reasons).

File: shiny/ui/_markdown_stream.py (hunk 15 above)
Comment (cpsievert) [reply shiny/ui/_markdown_stream.py:228]: If we call it `output_markdown_stream()`, would you expect there to be a `@render.markdown_stream`?

File: js/markdown-stream/markdown-stream.ts [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+    return;
+  }
+
+  if (message.operation === "replace") {
+    el.setAttribute("content", message.content);
+  } else if (message.operation === "append") {
+    const content = el.getAttribute("content");
+    el.setAttribute("content", content + message.content);
+  } else {
+    console.error(`Unknown operation: ${message.operation}`);
```
Comment (gadenbuie) [inline js/markdown-stream/markdown-stream.ts:249]: Might as well throw since it's in a message handler and won't bring down the app. That would just increase visibility (in dev mode we'd see the error in the client console).

File: shiny/ui/_markdown_stream.py (hunk 5 above)
Comment (gadenbuie) [reply shiny/ui/_markdown_stream.py]: lol I called this out in comments, I think the paired `@render...` function detail is probably not how most users will think of what's an output vs what's an input. I think that most people will see this as an "output" because you're sending something from the server and it's showing up in the UI.

File: shiny/ui/_markdown_stream.py (hunk 10 above)
Comment (gadenbuie) [reply shiny/ui/_markdown_stream.py]: Could we make it work? 

If not, I think the documentation needs to be a bit stronger: like _don't use this in Core!_ But I'd rather not solve this with documentation; if we can save people from self-inflicted issues we should do that. Even better if we can have a consistent interface that works the same way in both modes.

Comment (gadenbuie) [reply shiny/ui/_markdown_stream.py]: If we use `md.ui()`, we wouldn't have the "is it an output or not?" problem.

File: js/markdown-stream/markdown-stream.ts (hunk 16 above)
Comment (cpsievert) [reply js/markdown-stream/markdown-stream.ts]: Done in [add7b2f](https://github.com/posit-dev/py-shiny/pull/1782/commits/add7b2f48526adc67b8340ef77c2e76e8d4ac825)

File: shiny/ui/_markdown_stream.py (hunk 15 above)
Comment (gadenbuie) [reply shiny/ui/_markdown_stream.py:238]: Sorry, I found https://github.com/posit-dev/py-shiny/pull/1782#discussion_r1936243033 after writing this.

But no, from a user perspective I don't think I'd find it weird. For me, the defining feature of an output is that I'm sending something from the server and it's showing up in the UI; the `@render` part seems incidental, like that's just how it works for plots or tables.

File: js/markdown-stream/markdown-stream.ts (hunk 14 above)
Comment (cpsievert) [reply js/markdown-stream/markdown-stream.ts]: Done in [add7b2f](https://github.com/posit-dev/py-shiny/pull/1782/commits/add7b2f48526adc67b8340ef77c2e76e8d4ac825)

File: shiny/ui/_markdown_stream.py (hunk 12 above)
Comment (cpsievert) [reply shiny/ui/_markdown_stream.py:139]: Addressed as a note in [d091ebd](https://github.com/posit-dev/py-shiny/pull/1782/commits/d091ebdedb97b6f222f8284c5ac05695d1942b42)

If you feel strongly about `.append()`, would you mind filing an issue with some thoughts on it?

File: shiny/ui/_markdown_stream.py (hunk 11 above)
Comment (cpsievert) [reply shiny/ui/_markdown_stream.py]: Addressed in [d091ebd](https://github.com/posit-dev/py-shiny/pull/1782/commits/d091ebdedb97b6f222f8284c5ac05695d1942b42)

File: shiny/ui/_markdown_stream.py (hunk 7 above)
Comment (cpsievert) [reply shiny/ui/_markdown_stream.py]: Done in [d091ebd](https://github.com/posit-dev/py-shiny/pull/1782/commits/d091ebdedb97b6f222f8284c5ac05695d1942b42)

File: shiny/ui/_markdown_stream.py (hunk 13 above)
Comment (cpsievert) [reply shiny/ui/_markdown_stream.py:265]: > but we should also have content_type as an argument in stream()

I agree, but unfortunately it wouldn't be support that combined with `.stream(clear=False)` since you could end up with a mix of content types. Thats said, I guess we could just not allow `content_type` to be specified when `clear=False`?

I guess another way we could go to get around this is to have `.stream()` always clear/replace the content, and then a `.append_stream()` that wouldn't have `content_type`?

File: shiny/ui/_markdown_stream.py (hunk 10 above)
Comment (cpsievert) [reply shiny/ui/_markdown_stream.py]: I get where you're coming from. I think it could technically work since there's not currently any important state management going on, but it makes me feel uneasy that it would handcuff future design/engineering.

Also, the tone of the docs... 

> This method is only relevant for Shiny Express. In Shiny Core, use :func:`~shiny.ui.markdown_stream_ui` for placing the markdown stream in the UI.

... feels appropriate to me since there is no inherent danger in using `.ui()` in Core if you know what you're doing

File: shiny/ui/_markdown_stream.py (hunk 15 above)
Comment (cpsievert) [reply shiny/ui/_markdown_stream.py]: Ok, that's fair, done in [a99b3a5](https://github.com/posit-dev/py-shiny/pull/1782/commits/a99b3a5bc4a0cdbc403a897f4de08de8473a90be).

It does make me wonder though -- should we change `chat_ui()` to `output_chat()` (for similar reasons)? I remember going back and forth and that one too and can't quite remember why I landed on `chat_ui()` (probably because it just feels weird for something that's technically an output in classical sense)

File: shiny/ui/_markdown_stream.py (hunk 10 above)
Comment (gadenbuie) [reply shiny/ui/_markdown_stream.py]: I've been thinking about this while traveling and my opinion distills down to thinking that we should do whatever we can to make sure that the Core and Express usage and implementation are as unified as possible. If we can write a component such that you use it the same way in both syntaxes **and** the implementations are the same, it's worth the reduced complexity for future support.

The one shortcut I think we should allow ourselves is to automatically create UI elements when we can in Express – like not needing `ui.output_plot()` because it's obviated by `@render.plot`. For `MarkdownStream`, however, users still need to insert the UI in both cases. My objection is that how that UI is created is different between the two (instance method in Express, separate function in Core).

Another thing to consider is that the class-based approach is very natural in Python, but slightly at odds with how things work in Shiny in general. My sense is that many people want our components to work more like the Express version in this PR, but there are reasons why that's difficult. 

My feeling is that we should either figure out how to make a class-based component work all the way in all the places, or we should stick with an interface that's common to the rest of Shiny. Here's a patch that enables using the class-based approach in both Core and Express: 
[md-stream-core.patch](https://github.com/user-attachments/files/18643848/md-stream-core.patch). I'm not sure how this pattern will work out with modules, though.

Speaking of modules, I think the component needs to use `resolve_id()` throughout (added in the patch above) to support modules. But that got me thinking: maybe we could use Shiny modules for this component instead. That would look like this:

```python
ui = ui.page_fluid(
    ui.markdown_stream_ui("readme")
)

def server(input, output, session):
	md_readme = ui.markdown_stream_server("readme")

	# ... chunk generator code ...

    @reactive.effect
    async def _():
        await md_readme.stream(chunk_generator())
```

In other words, the server function would return the `MarkdownStream` instance, with the appropriate classes. In that case, because we're building on a syntax and pattern that already exists, we could considering having the express version of `markdown_stream_server()` emit the UI as well (or maybe just called `markdown_stream()` in Express.

```python
from shiny.express import ui

```python
md_readme = ui.markdown_stream_server("readme")

# ... chunk generator code ...

@reactive.effect
async def _():
    await md_readme.stream(chunk_generator())
```

File: shiny/ui/_markdown_stream.py (hunk 15 above)
Comment (gadenbuie) [reply shiny/ui/_markdown_stream.py]: > It does make me wonder though -- should we change `chat_ui()` to `output_chat()` (for similar reasons)?

I'd say the difference here is that Markdown Stream is unidirectional: it's only server → client. `chat_ui()` makes sense because there are also inputs: currently just the text input but it's very conceivable that we add additional input types to the UI. IOW, `chat_ui()` makes sense because it's not just an output.

**Comment by cpsievert** [comment]
Merging so this available to others at work week. Will file issues to handle remaining feedback

File: shiny/ui/_markdown_stream.py (hunk 10 above)
Comment (gadenbuie) [reply shiny/ui/_markdown_stream.py]: Summarized in https://github.com/posit-dev/py-shiny/issues/1837
